### PR TITLE
Restructure dev mode handler logic to avoid starting webpack twice

### LINKED ISSUE #8981: Two webpack-dev-servers started if the application is restarted during webpack startup
### Minimal reproducible example
```
vaadin init double-webpack --latest
cd double-webpack
mvn install # Only to generate webpack.config.js
```

Add the snippet to the start of webpack.config.js
```
const fs = require("fs");
if (!fs.existsSync("target/classes/foo.txt")) {
  fs.writeFileSync("target/classes/foo.txt","Hello");
}
```

Run 
```
mvn
```

### Expected behavior
One `webpack-dev-server` process is running

### Actual behavior
Two `webpack-dev-server` processes are running

### Versions:
    - Vaadin version: 17
    - Java version: 11
    - OS version: macOS

### PR DESCRIPTION
**Pull request by denis-anisimov**
Fixes #8981

### CODE REVIEW

File: flow-server/src/main/java/com/vaadin/flow/server/DevModeHandler.java [hunk 1]
```diff
@@ -188,8 +191,10 @@ public static DevModeHandler start(int runningPort,
                 || !configuration.enableDevServer()) {
             return null;
         }
-        atomicHandler.compareAndSet(null,
-                createInstance(runningPort, configuration, npmFolder, waitFor));
+        if (atomicHandler.get() == null) {
+            atomicHandler.compareAndSet(null, createInstance(runningPort,
```
Comment (joheriks) [inline flow-server/src/main/java/com/vaadin/flow/server/DevModeHandler.java:195]: Minor: Because of the added null check (I assume to avoid calling `createInstance` in not null case as there is no lazy `compareAndSet` available), could the inner code be simplified to `atomicHandler.set(...)`

**Review by joheriks — APPROVED** [review]

File: flow-server/src/main/java/com/vaadin/flow/server/DevModeHandler.java (hunk 1 above)
Comment (denis-anisimov) [reply flow-server/src/main/java/com/vaadin/flow/server/DevModeHandler.java:195]: No, it can't be.
That's the point of thread safety.
The whole idea of using AtomicReference is to update the value safely and avoid extra action.
If two threads comes here then both threads may pass `atomicHandler.get() == null`  check.
Then one thread may set it first and another thread will also create a new instance and set it but it should be rejected sine there is already one.
The added check will work in many cases to avoid even `createInstance`   call . But in the worst case it will be called anyway twice (or even more).
The changes which I made inside `DevModeHandler`  logic allows to avoid side effects in this case.
But `atomicHandler.get() == null`  check already will work in most cases so there will be no even need to create another instance.
So in short : the added check is not necessary (and should not be necessary) . But it allows to avoid extra actions in many cases even though it's not a booletproof solution.

Comment (joheriks) [reply flow-server/src/main/java/com/vaadin/flow/server/DevModeHandler.java:195]: Ok. I was under the impression that the `AtomicReference` is there to have thread safe access between calls to `DevModeHandler::start` and `DevModeHandler::stop`. Is there at all a possibility to have two simulatenous threads entering `DevModeHandler::start`?

Anyway, this is fine and yes probably safer.

Comment (denis-anisimov) [reply flow-server/src/main/java/com/vaadin/flow/server/DevModeHandler.java:195]: Theoretically this method can be invoked from two different threads.
Practically: I don't know. This is not my code. So I don't know all the details behind decision to make it thread safe.
But this way it's thread safe without side effects.

**Review by joheriks — APPROVED** [review]

**Comment by taefi** [comment]
Cherry-picked to `4.0` by @joheriks on 5.10.2020
